{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c26f321c",
   "metadata": {},
   "source": [
    "## Notebook for prepping the ADRD Aging genotypes for NIMH Human Brain Collection Core\n",
    "original 3 Illumina chip types ped/map files from [here](gs://nihnialng-aging-brain/genotypes); H1M, H5M4, M650K\n",
    "\n",
    "- H1M=Human1M-Duov3_B\n",
    "- H650K=HumanHap650Yv3.0\n",
    "- H5M4=HumanOmni5-Quad\n",
    "\n",
    "basically need to:\n",
    "- merge genotypes from different Illumina platforms, have two sets of these one from phase1 and seperate set for just the additional phase2 subjects\n",
    "- liftover from hg19 to hg38\n",
    "- re-order the chromosomes from typical to 10X's lexigraphical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2d9f1fe9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Aug 11 16:02:18 EDT 2023\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc23a89e",
   "metadata": {},
   "source": [
    "#### import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1c144911",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pandas import read_csv, DataFrame\n",
    "from os.path import exists\n",
    "from os import sched_getaffinity\n",
    "from numpy import append"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91ecc116-f7fc-41f9-ba53-965821d1e703",
   "metadata": {},
   "source": [
    "#### set notebook variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "17c2f586",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# naming\n",
    "proj_name = 'aging_phase2'\n",
    "\n",
    "# directories\n",
    "wrk_dir = '/labshare/raph/datasets/adrd_neuro/brain_aging/phase2'\n",
    "genos_dir = f'{wrk_dir}/genotypes'\n",
    "info_dir = f'{wrk_dir}/sample_info'\n",
    "public_dir = f'{wrk_dir}/public'\n",
    "src_dir = f'{wrk_dir}/src_data'\n",
    "\n",
    "# in files\n",
    "info_file = f'{info_dir}/{proj_name}.sample_info.csv'\n",
    "demuxlet_vcf_file = f'{genos_dir}/{proj_name}.hg38.demuxlet.vcf.gz'\n",
    "fasta_index_bucket_path = 'gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta*'\n",
    "fasta_dict_bucket_path = 'gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dict'\n",
    "\n",
    "# variables\n",
    "DEBUG = False\n",
    "cpu_count = len(sched_getaffinity(0))\n",
    "capture_out = !grep MemTotal /proc/meminfo | awk '{print $2}'\n",
    "max_mem = int(capture_out[0])\n",
    "\n",
    "hbcc_ped_prefixes = ['H1M', 'H5M4', 'H650K']\n",
    "# when combining final vcf for demuxlet need to have chromosomes \n",
    "#sorted to match 10x lexigraphical\n",
    "autosomes = [1, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 2, 20, 21, 22, 3, 4, 5, 6, 7, 8, 9]\n",
    "lane_range = range(1, 9)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13d42fb3-47c9-4d79-a389-e7b02a42a4c8",
   "metadata": {},
   "source": [
    "#### utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7b173107-efef-4b40-96e0-e121d0bf6ea9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def run_bash_command(cmd_line: str, verbose: bool=False):\n",
    "    if verbose:\n",
    "        print(cmd_line)\n",
    "    !{cmd_line}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8478abcd-3da3-41c2-aed0-940f0577b83d",
   "metadata": {},
   "source": [
    "### load sample info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f1f1992a-4840-420f-b8b8-01fecb011124",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of sample info: (36, 13)\n"
     ]
    }
   ],
   "source": [
    "info_df = read_csv(info_file)\n",
    "print(f'shape of sample info: {info_df.shape}')\n",
    "if DEBUG:\n",
    "    display(info_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03d984d6-9341-4b42-b03d-179fff392839",
   "metadata": {},
   "source": [
    "#### drop the non-pooled samples and make sure pool nums are ints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3f24d23c-65c3-4f8a-b413-70ed98048ac0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of info (33, 13)\n"
     ]
    }
   ],
   "source": [
    "info_df = info_df.loc[(~info_df.gex_pool.isna()) & (~info_df.atac_pool.isna())]\n",
    "print(f'shape of info {info_df.shape}')\n",
    "# make sure pool nums are ints and not floats\n",
    "info_df.gex_pool = info_df.gex_pool.astype('int')\n",
    "info_df.atac_pool = info_df.atac_pool.astype('int')\n",
    "if DEBUG:\n",
    "    display(info_df.head())\n",
    "    display(info_df.gex_pool.value_counts())\n",
    "    display(info_df.atac_pool.value_counts())    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d9e2225-86be-480a-8e05-4b73dcaf9dec",
   "metadata": {},
   "source": [
    "### convert pedmaps to bfiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "34f59700",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 212 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/phase1/aging_phase2_H1M.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: 3126 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/phase1/aging_phase2_H5M4.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: 103 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/phase2/aging_phase2_H1M.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: 9849 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/phase2/aging_phase2_H5M4.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n"
     ]
    }
   ],
   "source": [
    "for phase in ['phase1', 'phase2']:\n",
    "    for chip_type in hbcc_ped_prefixes:\n",
    "        this_cmd = f'plink --ped {genos_dir}/{phase}/{chip_type}.ped \\\n",
    "--map {genos_dir}/{phase}/{chip_type}.map \\\n",
    "--make-bed --out {genos_dir}/{phase}/{proj_name}_{chip_type} --silent'\n",
    "        run_bash_command(this_cmd, DEBUG)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "589e10c7-841d-404c-969a-b2aee667b676",
   "metadata": {},
   "source": [
    "### merge the project phases per geno chip platform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4b59813a-c370-4f29-91fb-a1973441543e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H1M\n",
      "Warning: Variants 'rs5746074' and 'GA018547' have the same position.\n",
      "Warning: Variants 'GA000381' and '200240' have the same position.\n",
      "Warning: Variants 'GA018659' and '200251' have the same position.\n",
      "162 more same-position warnings: see log file.\n",
      "H5M4\n",
      "Warning: Variants 'rs17160939' and 'kgp15219107' have the same position.\n",
      "Warning: Variants 'rs6694982' and 'kgp15412548' have the same position.\n",
      "Warning: Variants 'kgp2466911' and 'GA018352' have the same position.\n",
      "73814 more same-position warnings: see log file.\n",
      "H650K\n"
     ]
    }
   ],
   "source": [
    "for chip_type in hbcc_ped_prefixes:\n",
    "    print(chip_type)\n",
    "    this_cmd = f'plink --bfile {genos_dir}/phase1/{proj_name}_{chip_type} \\\n",
    "--bmerge {genos_dir}/phase2/{proj_name}_{chip_type} \\\n",
    "--out {genos_dir}/{proj_name}_{chip_type} --silent'\n",
    "    run_bash_command(this_cmd, DEBUG)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a78067bf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 284M\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 4.6M Aug 11 16:02 aging_phase2_H1M.bed\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  39M Aug 11 16:02 aging_phase2_H1M.bim\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  402 Aug 11 16:02 aging_phase2_H1M.fam\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  13K Aug 11 16:02 aging_phase2_H1M.log\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  13M Aug 11 16:02 aging_phase2_H5M4.bed\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 146M Aug 11 16:02 aging_phase2_H5M4.bim\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  378 Aug 11 16:02 aging_phase2_H5M4.fam\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 5.1M Aug 11 16:02 aging_phase2_H5M4.log\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 1.9M Aug 11 16:02 aging_phase2_H650K.bed\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  22M Aug 11 16:02 aging_phase2_H650K.bim\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  252 Aug 11 16:02 aging_phase2_H650K.fam\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 1.6K Aug 11 16:02 aging_phase2_H650K.log\n",
      "drwxrwxr-x. 2 gibbsr gibbsr 4.0K Aug 11 16:02 phase1\n",
      "drwxrwxr-x. 2 gibbsr gibbsr 8.0K Aug 11 16:02 phase2\n"
     ]
    }
   ],
   "source": [
    "!ls -lh {genos_dir}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b5f48b9",
   "metadata": {},
   "source": [
    "#### see what variants are shared between platforms provided, by variant name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fc47125d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H1M shape=(1192666, 6) shared variant size = 1192666\n",
      "H5M4 shape=(4437269, 6) shared variant size = 971362\n",
      "H650K shape=(660918, 6) shared variant size = 561035\n"
     ]
    }
   ],
   "source": [
    "shared_variants = None\n",
    "for chip_type in hbcc_ped_prefixes:\n",
    "    bim_df = read_csv(f'{genos_dir}/{proj_name}_{chip_type}.bim', sep='\\s+', header=None)\n",
    "    # in map file variants name is 2nd column\n",
    "    if shared_variants is None:\n",
    "        shared_variants = set(bim_df[1])\n",
    "    else:\n",
    "        shared_variants = shared_variants & set(bim_df[1])\n",
    "    print(f'{chip_type} shape={bim_df.shape} shared variant size = {len(shared_variants)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b0a72cc",
   "metadata": {},
   "source": [
    "#### looks like a definitely a decent number of variants shared, proceed to merging"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "376bf1a9",
   "metadata": {},
   "source": [
    "#### merge the per Illumina chip type for the plink bfiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "71d455e9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Multiple chromosomes seen for variant 'rs10402666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1049702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1131675'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11551867'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11799112'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11925488'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1224751'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1640283'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17883004'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1851319'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1864011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2005417'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2090052'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs252585'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2555849'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2622254'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2635512'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs281377'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28539253'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28693514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2957297'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3212680'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34278206'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35437602'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs381494'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4797621'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs496572'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5945676'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6830714'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs685438'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7212153'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7255768'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs731137'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7600058'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8031106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9647005'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9778500'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9785726'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786108'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786272'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA018897'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7530935'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10182914'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9752952'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA020826'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA022478'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs909274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13154594'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10038870'.\n",
      "Warning: Multiple positions seen for variant 'rs10478694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10807011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35882598'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10260309'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2261257'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3915411'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34499424'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28580083'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA027818'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA011118'.\n",
      "Warning: Multiple positions seen for variant 'rs35661597'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17212015'.\n",
      "Warning: Multiple positions seen for variant 'rs9561064'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs380804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12597978'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2690904'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7267880'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821958'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2334639'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs140281'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13056641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5752647'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2771839'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2568759'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35858670'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34730793'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3102685'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2562942'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2163209'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9782855'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786212'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35740538'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5941215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2552960'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7885800'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2481225'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2437650'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35324435'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34262498'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9657839'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2556924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2565525'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2751196'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4028872'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2435738'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5985048'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2933080'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1774963'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17129789'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs315041'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4364877'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17101527'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2655760'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4649592'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12043679'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1079109'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12128686'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6427609'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2376852'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2754471'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2351422'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2348472'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2339752'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12619064'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2176413'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2192400'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6546667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10200219'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6710843'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12615215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1498957'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6421783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2375763'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs163667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1519514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13425421'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13002544'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2342694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11717239'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9871790'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2028240'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7624950'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918537'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918531'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11706032'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs536575'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3009033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918515'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13326562'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9871837'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1568292'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7628649'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2895461'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2063300'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2279029'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3860706'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1983217'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2291715'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9761692'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2662733'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3925117'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4276244'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10015934'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13120250'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11738302'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13154594'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3903000'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12523271'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11738629'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2569201'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1826532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4959515'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs719065'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596790'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596796'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1535053'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12660307'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2158556'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6937363'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9328044'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2281351'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17133610'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10155688'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1316662'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596815'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9501714'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12191783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6915646'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12192106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6916246'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6913464'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9502915'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9405480'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9502941'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1080100'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1610415'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1045074'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9264536'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs602875'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12660719'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040035'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11760174'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12183238'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1112237'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11759626'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2056474'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10245070'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10224134'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2257576'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs38642'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17152154'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2456554'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17863175'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361379'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163396'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361342'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2252241'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163403'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2213198'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2213199'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13233002'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361377'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11535222'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855945'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2734189'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17282'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16622'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17708955'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855950'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855951'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2734151'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855150'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855954'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12672113'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6947359'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10487532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12534123'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13223129'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1006094'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6979080'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12533751'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1006093'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17257'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17766702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163497'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361473'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7787291'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17767132'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2269332'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17286'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10231034'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12667623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10215320'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040350'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040351'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7780065'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7789642'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855983'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2680432'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1126092'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs682400'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2035685'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs974053'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13258887'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2380567'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2338696'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1113706'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1126274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1121947'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4976940'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7015131'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6421025'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11786011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1446931'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1446934'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7819587'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1470514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11781564'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13291255'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1112042'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10908023'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1578263'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10123966'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2282199'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13285529'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs715666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2026726'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7041980'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11103281'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7025716'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4842084'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs884559'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4842090'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1073731'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1751274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs564166'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10508547'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2437258'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs691461'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2461180'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2813267'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2394498'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10881601'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1158744'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10836370'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2901810'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs714215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4244818'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11820113'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3855349'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10502267'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1975930'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs770987'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1296265'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1278610'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7131737'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11611532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs710944'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7306551'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs770991'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758919'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2272367'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2036288'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7300570'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12422489'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758929'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11611755'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7304881'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7954265'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7979175'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3825108'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7977661'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3847667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7310221'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1587447'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6560914'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758928'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7970251'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7961667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7961682'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs905227'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3805728'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7958575'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11297'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2259444'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11612324'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs595241'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3749826'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2088683'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1924627'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2897385'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2038897'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7984879'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6492347'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11159741'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1954089'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12442023'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2665033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12591365'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2871771'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs438702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1685119'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2858016'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1104612'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4985430'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4788728'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1609516'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1973344'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9708471'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs883635'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs900768'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9284357'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1079841'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3907609'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4636966'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3999623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12939215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4968723'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8069746'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8069177'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4968624'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2070782'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11079538'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7212440'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8066263'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9912957'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8065316'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7503015'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7503902'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9907512'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8080666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7210806'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9899114'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9894547'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7501783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7405659'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8071835'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9889615'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9896532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1619323'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2318319'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs918442'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10414884'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3897509'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8130901'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3855691'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821973'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821997'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2792379'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2775862'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17004452'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2258197'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2822124'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs743067'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1985740'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2258691'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2775054'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11700583'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17001860'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17728665'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361973'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2845371'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5993924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10084637'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2190742'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748614'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748622'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9605145'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9605146'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs759235'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2108585'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16981694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9606468'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748636'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7290901'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5752043'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11912332'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13056641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16979433'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5760804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7286373'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11705626'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs713696'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8142181'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5760813'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs742169'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5996821'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6004408'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2064139'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6005641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1883472'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6519730'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs882455'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs882456'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820778'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6005657'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1004236'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs722343'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11090516'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1062731'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1997739'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs730647'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs14115'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2413140'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5754192'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4821082'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820087'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788477'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12166565'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788479'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5994595'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740985'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740987'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740991'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs114682'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740982'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs130743'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs470033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820090'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1159220'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5998577'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4821086'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788483'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs713867'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs136183'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3021268'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11545763'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2061106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5918554'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6619104'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12557859'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6418513'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6627746'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5925325'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7058908'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5924804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2097253'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17254754'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC464T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA829G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC2791T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC3595T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT4562C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT6153C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA6753G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoG8270A'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT8278C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC8656T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT11486C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT14799C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA15245G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoG16392A'.\n",
      "653506 more multiple-position warnings: see log file.\n",
      "Error: 1 variant with 3+ alleles present.\n",
      "* If you believe this is due to strand inconsistency, try --flip with\n",
      "  /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2-merge.missnp.\n",
      "  (Warning: if this seems to work, strand errors involving SNPs with A/T or C/G\n",
      "  alleles probably remain in your data.  If LD between nearby SNPs is high,\n",
      "  --flip-scan should detect them.)\n",
      "* If you are dealing with genuine multiallelic variants, we recommend exporting\n",
      "  that subset of the data to VCF (via e.g. '--recode vcf'), merging with\n",
      "  another tool/script, and then importing the result; PLINK is not yet suited\n",
      "  to handling them.\n",
      "See https://www.cog-genomics.org/plink/1.9/data#merge3 for more discussion.\n",
      "Warning: 315 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2_H1M.temp.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: 12975 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2_H5M4.temp.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10402666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1049702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1131675'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11551867'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11799112'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11925488'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1224751'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1640283'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17883004'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1851319'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1864011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2005417'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2090052'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs252585'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2555849'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2622254'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2635512'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs281377'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28539253'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28693514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2957297'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3212680'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34278206'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35437602'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs381494'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4797621'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs496572'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5945676'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6830714'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs685438'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7212153'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7255768'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs731137'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7600058'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8031106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9647005'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9778500'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9785726'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786108'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786272'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA018897'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7530935'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10182914'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9752952'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA020826'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA022478'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs909274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13154594'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10038870'.\n",
      "Warning: Multiple positions seen for variant 'rs10478694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10807011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35882598'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10260309'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2261257'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3915411'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34499424'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28580083'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA027818'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA011118'.\n",
      "Warning: Multiple positions seen for variant 'rs35661597'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17212015'.\n",
      "Warning: Multiple positions seen for variant 'rs9561064'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs380804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12597978'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2690904'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7267880'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821958'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2334639'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs140281'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13056641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5752647'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2771839'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2568759'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35858670'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34730793'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3102685'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2562942'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2163209'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9782855'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786212'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35740538'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5941215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2552960'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7885800'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2481225'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2437650'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35324435'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34262498'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9657839'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2556924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2565525'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2751196'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4028872'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2435738'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5985048'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2933080'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1774963'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17129789'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs315041'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4364877'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17101527'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2655760'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4649592'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12043679'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1079109'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12128686'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6427609'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2376852'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2754471'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2351422'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2348472'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2339752'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12619064'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2176413'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2192400'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6546667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10200219'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6710843'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12615215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1498957'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6421783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2375763'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs163667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1519514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13425421'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13002544'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2342694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11717239'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9871790'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2028240'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7624950'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918537'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918531'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11706032'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs536575'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3009033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918515'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13326562'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9871837'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1568292'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7628649'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2895461'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2063300'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2279029'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3860706'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1983217'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2291715'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9761692'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2662733'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3925117'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4276244'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10015934'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13120250'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11738302'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13154594'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3903000'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12523271'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11738629'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2569201'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1826532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4959515'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs719065'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596790'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596796'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1535053'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12660307'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2158556'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6937363'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9328044'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2281351'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17133610'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10155688'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1316662'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596815'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9501714'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12191783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6915646'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12192106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6916246'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6913464'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9502915'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9405480'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9502941'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1080100'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1610415'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1045074'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9264536'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs602875'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12660719'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040035'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11760174'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12183238'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1112237'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11759626'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2056474'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10245070'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10224134'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2257576'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs38642'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17152154'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2456554'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17863175'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361379'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163396'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361342'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2252241'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163403'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2213198'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2213199'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13233002'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361377'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11535222'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855945'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2734189'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17282'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16622'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17708955'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855950'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855951'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2734151'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855150'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855954'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12672113'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6947359'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10487532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12534123'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13223129'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1006094'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6979080'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12533751'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1006093'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17257'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17766702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163497'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361473'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7787291'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17767132'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2269332'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17286'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10231034'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12667623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10215320'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040350'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040351'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7780065'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7789642'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855983'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2680432'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1126092'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs682400'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2035685'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs974053'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13258887'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2380567'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2338696'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1113706'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1126274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1121947'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4976940'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7015131'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6421025'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11786011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1446931'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1446934'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7819587'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1470514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11781564'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13291255'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1112042'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10908023'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1578263'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10123966'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2282199'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13285529'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs715666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2026726'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7041980'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11103281'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7025716'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4842084'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs884559'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4842090'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1073731'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1751274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs564166'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10508547'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2437258'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs691461'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2461180'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2813267'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2394498'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10881601'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1158744'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10836370'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2901810'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs714215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4244818'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11820113'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3855349'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10502267'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1975930'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs770987'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1296265'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1278610'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7131737'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11611532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs710944'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7306551'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs770991'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758919'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2272367'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2036288'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7300570'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12422489'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758929'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11611755'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7304881'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7954265'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7979175'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3825108'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7977661'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3847667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7310221'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1587447'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6560914'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758928'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7970251'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7961667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7961682'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs905227'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3805728'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7958575'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11297'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2259444'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11612324'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs595241'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3749826'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2088683'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1924627'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2897385'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2038897'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7984879'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6492347'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11159741'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1954089'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12442023'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2665033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12591365'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2871771'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs438702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1685119'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2858016'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1104612'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4985430'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4788728'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1609516'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1973344'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9708471'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs883635'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs900768'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9284357'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1079841'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3907609'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4636966'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3999623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12939215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4968723'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8069746'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8069177'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4968624'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2070782'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11079538'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7212440'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8066263'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9912957'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8065316'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7503015'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7503902'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9907512'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8080666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7210806'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9899114'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9894547'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7501783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7405659'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8071835'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9889615'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9896532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1619323'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2318319'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs918442'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10414884'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3897509'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8130901'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3855691'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821973'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821997'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2792379'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2775862'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17004452'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2258197'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2822124'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs743067'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1985740'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2258691'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2775054'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11700583'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17001860'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17728665'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361973'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2845371'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5993924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10084637'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2190742'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748614'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748622'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9605145'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9605146'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs759235'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2108585'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16981694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9606468'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748636'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7290901'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5752043'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11912332'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13056641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16979433'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5760804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7286373'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11705626'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs713696'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8142181'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5760813'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs742169'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5996821'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6004408'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2064139'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6005641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1883472'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6519730'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs882455'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs882456'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820778'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6005657'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1004236'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs722343'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11090516'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1062731'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1997739'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs730647'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs14115'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2413140'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5754192'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4821082'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820087'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788477'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12166565'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788479'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5994595'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740985'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740987'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740991'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs114682'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740982'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs130743'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs470033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820090'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1159220'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5998577'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4821086'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788483'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs713867'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs136183'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3021268'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11545763'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2061106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5918554'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6619104'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12557859'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6418513'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6627746'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5925325'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7058908'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5924804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2097253'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17254754'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC464T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA829G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC2791T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC3595T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT4562C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT6153C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA6753G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoG8270A'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT8278C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC8656T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT11486C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT14799C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA15245G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoG16392A'.\n",
      "653505 more multiple-position warnings: see log file.\n",
      "Warning: Variants 'rs4040617' and 'kgp5225889' have the same position.\n",
      "Warning: Variants 'rs17160939' and 'kgp15219107' have the same position.\n",
      "Warning: Variants 'rs28609852' and 'kgp3324955' have the same position.\n",
      "207087 more same-position warnings: see log file.\n",
      "Warning: 13284 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n"
     ]
    }
   ],
   "source": [
    "# merge the files into a single plink binary set\n",
    "merge_file_set = f'{genos_dir}/merge-list.txt'\n",
    "bfile_set = f'{genos_dir}/{proj_name}'\n",
    "\n",
    "with open(merge_file_set, 'w') as file_handler:\n",
    "    for chip_type in hbcc_ped_prefixes:\n",
    "        prefix_file_set = f'{genos_dir}/{proj_name}_{chip_type}'\n",
    "        file_handler.write(f'{prefix_file_set}\\n')\n",
    "\n",
    "# merge the per platform bfiles into a merged bfile\n",
    "this_cmd = f'plink --merge-list {merge_file_set} --make-bed --allow-no-sex \\\n",
    "--keep-allele-order --silent --out {bfile_set}'\n",
    "run_bash_command(this_cmd, DEBUG)\n",
    "\n",
    "# if there was a missnp problem remove those variant and re-attemp merge\n",
    "if exists(f'{bfile_set}-merge.missnp'):\n",
    "    for chip_type in hbcc_ped_prefixes:\n",
    "        this_cmd = f'plink --bfile {genos_dir}/{proj_name}_{chip_type} \\\n",
    "--silent --exclude {bfile_set}-merge.missnp \\\n",
    "--keep-allele-order --make-bed \\\n",
    "--out {genos_dir}/{proj_name}_{chip_type}.temp'\n",
    "        run_bash_command(this_cmd, DEBUG)\n",
    "\n",
    "    with open(merge_file_set, 'w') as file_handler:\n",
    "        for chip_type in hbcc_ped_prefixes:\n",
    "            prefix_file_set = f'{genos_dir}/{proj_name}_{chip_type}.temp'\n",
    "            file_handler.write(f'{prefix_file_set}\\n')\n",
    "        \n",
    "    this_cmd = f'plink --merge-list {merge_file_set} --make-bed --allow-no-sex \\\n",
    "--keep-allele-order --silent --out {bfile_set} --geno 0.05'\n",
    "    run_bash_command(this_cmd, DEBUG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a87243ee",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r--. 1 gibbsr gibbsr  162 Aug 11 16:03 /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.nosex\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 446K Aug 11 16:03 /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hh\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 4.5M Aug 11 16:03 /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.bed\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 1.1K Aug 11 16:03 /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.fam\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  17M Aug 11 16:03 /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.bim\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  51M Aug 11 16:03 /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.log\n"
     ]
    }
   ],
   "source": [
    "!ls -lhtr {genos_dir}/{proj_name}.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "de70d3a3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35 people (15 males, 10 females, 10 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.nosex\n",
      ".\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 35 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... done.\n",
      "Warning: 13284 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.446112.\n",
      "4141901 variants removed due to missing genotype data (--geno).\n",
      "523522 variants and 35 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.bed\n",
      "+\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.bim\n",
      "+\n",
      "/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.fam\n",
      "... done.\n",
      "\n",
      "End time: Fri Aug 11 16:03:32 2023\n"
     ]
    }
   ],
   "source": [
    "!tail -n 25 {genos_dir}/{proj_name}.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2e8d51a",
   "metadata": {},
   "source": [
    "#### load the bim files see chrom counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d6ebcb4f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of bim file (523522, 6)\n"
     ]
    }
   ],
   "source": [
    "bim_df = read_csv(f'{bfile_set}.bim', sep='\\s+', header=None)\n",
    "print(f'shape of bim file {bim_df.shape}')\n",
    "if DEBUG:\n",
    "    display(bim_df.sample(10))\n",
    "    display(bim_df[0].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8879453",
   "metadata": {},
   "source": [
    "#### find variants to exclude that aren't SNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "902b9543",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A    278976\n",
       "G    193036\n",
       "C     45887\n",
       "0      5619\n",
       "T         4\n",
       "Name: 4, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "A    241467\n",
       "G    230607\n",
       "C     51430\n",
       "T        18\n",
       "Name: 5, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vars to include shape (517903, 6)\n"
     ]
    }
   ],
   "source": [
    "display(bim_df[4].value_counts())\n",
    "display(bim_df[5].value_counts())\n",
    "nucleotides = ['A', 'C', 'G', 'T', 'N', 'a', 'c', 'g', 't', 'n', 0]\n",
    "vars_to_include = bim_df.loc[(bim_df[4].isin(nucleotides)) & \n",
    "                             (bim_df[5].isin(nucleotides))]\n",
    "print(f'vars to include shape {vars_to_include.shape}')\n",
    "if DEBUG:\n",
    "    display(vars_to_include.head())\n",
    "\n",
    "vars_to_include[1].to_csv(f'{genos_dir}/variants_to_keep.txt', index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17d8f947",
   "metadata": {},
   "source": [
    "### convert to vcf, exclude InDels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "90892b07",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "this_cmd = f'plink2 --bfile {bfile_set} --silent \\\n",
    "--export vcf-4.2 bgz id-paste=iid --out {bfile_set} \\\n",
    "--output-chr chrM --not-chr 0 --snps-only \\\n",
    "--extract {genos_dir}/variants_to_keep.txt'\n",
    "\n",
    "run_bash_command(this_cmd, DEBUG)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "930d8545",
   "metadata": {},
   "source": [
    "### re-ID the NIMH HBCC samples using the provided mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f34709b1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# create a rename file for bcftools to use\n",
    "temp_df = info_df.loc[~info_df.geno_IID.isna()]\n",
    "temp_df[['geno_IID', 'sample_id']].to_csv(f'{bfile_set}.rename.sample.list', \n",
    "                                        index=False, header=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2b9cf465",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "in_vcf = f'{bfile_set}.vcf.gz'\n",
    "out_vcf = f'{bfile_set}.renamed.vcf.gz'\n",
    "this_cmd = f'bcftools reheader --sample {bfile_set}.rename.sample.list \\\n",
    "--output {out_vcf} --threads 2 {in_vcf}'\n",
    "run_bash_command(this_cmd, DEBUG)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b02b6110-2564-4d03-a279-1855357b388a",
   "metadata": {},
   "source": [
    "#### check which subjects are present and properly renamed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e1ccaf10-80e1-4eef-b2a1-1a52032a166c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of IDs present (35, 1)\n",
      "expected but not found {'Aging134', 'Aging142'}\n",
      "found but not expected {'4040296074_A', '3999495136_R02C01', '4463344122_R01C02', '4572348740_R01C02'}\n"
     ]
    }
   ],
   "source": [
    "temp_sample_list_file = f'{bfile_set}.sample.list'\n",
    "this_cmd = f'bcftools query --list-samples {out_vcf} > {temp_sample_list_file}'\n",
    "run_bash_command(this_cmd, DEBUG)\n",
    "ids_present_df = read_csv(temp_sample_list_file, header=None)\n",
    "ids_present_df.columns = ['sample_id']\n",
    "print(f'shape of IDs present {ids_present_df.shape}')\n",
    "\n",
    "# which expected aren't present\n",
    "temp_result = set(info_df.sample_id) - set(ids_present_df.sample_id)\n",
    "print(f'expected but not found {temp_result}')\n",
    "# and then found but not exprected\n",
    "temp_result = set(ids_present_df.sample_id) - set(info_df.sample_id)\n",
    "print(f'found but not expected {temp_result}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3d7bd10",
   "metadata": {},
   "source": [
    "#### need to liftover from hg19 to hg38\n",
    "use Picard, slower but better than CrossMap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc40e055",
   "metadata": {},
   "source": [
    "#### get Picard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "dd995a0f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# grab picard jar\n",
    "!wget --quiet  https://github.com/broadinstitute/picard/releases/download/2.27.5/picard.jar \\\n",
    "-O {public_dir}/picard.jar"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64da3751",
   "metadata": {},
   "source": [
    "#### grab necessary ref files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "83636ddc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "this_cmd = f'gsutil -mq cp -P {fasta_index_bucket_path} {public_dir}/'\n",
    "run_bash_command(this_cmd, DEBUG)\n",
    "\n",
    "this_cmd = f'gsutil -mq cp -P {fasta_dict_bucket_path} {public_dir}/'\n",
    "run_bash_command(this_cmd, DEBUG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a202c6f0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# also need the hg19 files\n",
    "this_cmd = f'gsutil -mq cp -P gs://gcp-public-data--broad-references/hg19/v0/Homo_sapiens_assembly19.fasta* {public_dir}/'\n",
    "run_bash_command(this_cmd, DEBUG)\n",
    "\n",
    "this_cmd = f'gsutil -mq cp -P gs://gcp-public-data--broad-references/hg19/v0/Homo_sapiens_assembly19.dict {public_dir}/'\n",
    "run_bash_command(this_cmd, DEBUG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "666e50b7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!wget --quiet  http://hgdownload.cse.ucsc.edu/gbdb/hg19/liftOver/hg19ToHg38.over.chain.gz \\\n",
    "-O {public_dir}/hg19ToHg38.over.chain.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1a5c199",
   "metadata": {},
   "source": [
    "#### run the liftover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "71b3a25b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO\t2023-08-11 16:06:41\tLiftoverVcf\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    LiftoverVcf -INPUT /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.renamed.vcf.gz -OUTPUT /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.vcf.gz -CHAIN /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/public/hg19ToHg38.over.chain.gz -REJECT /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38unmapped.vcf.gz -REFERENCE_SEQUENCE /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/public/Homo_sapiens_assembly38.fasta -MAX_RECORDS_IN_RAM 5000000 -QUIET true -RECOVER_SWAPPED_REF_ALT true\n",
      "**********\n",
      "\n",
      "\n",
      "16:06:41.386 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/labshare/raph/datasets/adrd_neuro/brain_aging/phase2/public/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "INFO\t2023-08-11 16:06:41\tLiftoverVcf\tLoading up the target reference genome.\n",
      "INFO\t2023-08-11 16:06:47\tLiftoverVcf\tLifting variants over and sorting (not yet writing the output file.)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tProcessed 517598 variants.\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\t219 variants failed to liftover.\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\t258845 variants lifted over but had mismatching reference alleles after lift over.\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\t50.0512% of variants were not successfully lifted over and written to the output.\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tliftover success by source contig:\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr1: 19596 / 39394 (49.7436%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr10: 13625 / 27545 (49.4645%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr11: 12666 / 25271 (50.1207%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr12: 12613 / 25182 (50.0874%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr13: 10194 / 20149 (50.5931%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr14: 8435 / 16878 (49.9763%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr15: 7567 / 15274 (49.5417%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr16: 7686 / 15132 (50.7930%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr17: 6368 / 12667 (50.2724%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr18: 8014 / 16056 (49.9128%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr19: 3842 / 7751 (49.5678%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr2: 20908 / 41935 (49.8581%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr20: 6603 / 13131 (50.2856%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr21: 3770 / 7428 (50.7539%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr22: 3406 / 6745 (50.4967%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr3: 17587 / 35045 (50.1840%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr4: 15463 / 31189 (49.5784%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr5: 16045 / 31992 (50.1532%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr6: 16886 / 33771 (50.0015%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr7: 13719 / 27822 (49.3099%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr8: 14784 / 29449 (50.2020%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr9: 12077 / 24327 (49.6444%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchrM: 24 / 42 (57.1429%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchrX: 6448 / 13000 (49.6000%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchrXY: 0 / 13 (0.0000%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchrY: 208 / 410 (50.7317%)\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tlifted variants by target contig:\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr1: 19596\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr10: 13625\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr11: 12666\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr12: 12613\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr13: 10194\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr14: 8435\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr15: 7567\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr16: 7686\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr17: 6368\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr18: 8014\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr19: 3842\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr2: 20908\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr20: 6603\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr21: 3770\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr22: 3406\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr3: 17587\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr4: 15463\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr5: 16045\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr6: 16886\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr7: 13690\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr7_KI270803v1_alt: 29\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr8: 14775\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr8_KI270821v1_alt: 9\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchr9: 12077\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchrM: 24\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchrX: 6448\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tchrY: 208\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\t76200 variants were lifted by swapping REF/ALT alleles.\n",
      "INFO\t2023-08-11 16:06:58\tLiftoverVcf\tWriting out sorted records to final VCF.\n"
     ]
    }
   ],
   "source": [
    "vcf = f'{bfile_set}.renamed.vcf.gz'\n",
    "out_vcf_name = f'{bfile_set}.hg38.vcf.gz'\n",
    "out_vcf_unmapped_name = f'{bfile_set}.hg38unmapped.vcf.gz'\n",
    "\n",
    "this_cmd = \tf'java -Xmx{max_mem}k -jar {public_dir}/picard.jar LiftoverVcf \\\n",
    "INPUT={vcf} \\\n",
    "OUTPUT={out_vcf_name} \\\n",
    "CHAIN={public_dir}/hg19ToHg38.over.chain.gz \\\n",
    "REJECT={out_vcf_unmapped_name} \\\n",
    "REFERENCE_SEQUENCE={public_dir}/Homo_sapiens_assembly38.fasta \\\n",
    "MAX_RECORDS_IN_RAM=5000000 QUIET=true RECOVER_SWAPPED_REF_ALT=true'\n",
    "\n",
    "run_bash_command(this_cmd, DEBUG)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b35486a",
   "metadata": {},
   "source": [
    "#### split vcf by chromosome so can be recombined in specified chromosome order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cb54ba87",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.10.11.12.13.14.15.16.17.18.19.2.20.21.22.3.4.5.6.7.8.9.CPU times: user 67 ms, sys: 250 ms, total: 317 ms\n",
      "Wall time: 5.72 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "vcf_files = []\n",
    "for chrom in autosomes:\n",
    "    in_vcf = f'{bfile_set}.hg38.vcf.gz'\n",
    "    out_vcf = f'{bfile_set}.hg38.chr{chrom}'\n",
    "    # use plink2 instead of bcftools so header is reduced specifically for contigs\n",
    "    this_cmd = f'plink2 --vcf {in_vcf} --silent --chr {chrom} --not-chr 0 \\\n",
    "--export vcf-4.2 bgz id-paste=iid --out {out_vcf} --output-chr chrM --allow-extra-chr'\n",
    "    print(chrom, end='.')\n",
    "    run_bash_command(this_cmd, DEBUG)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0f8bddb",
   "metadata": {},
   "source": [
    "#### concat chromosome vcfs (in order) into genome vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "dd78507c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking the headers and starting positions of 22 files\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr1.vcf.gz\t0.023344 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr10.vcf.gz\t0.009232 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr11.vcf.gz\t0.006755 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr12.vcf.gz\t0.005781 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr13.vcf.gz\t0.005124 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr14.vcf.gz\t0.006616 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr15.vcf.gz\t0.004019 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr16.vcf.gz\t0.003830 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr17.vcf.gz\t0.004005 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr18.vcf.gz\t0.004298 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr19.vcf.gz\t0.003716 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr2.vcf.gz\t0.009719 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr20.vcf.gz\t0.003814 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr21.vcf.gz\t0.003062 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr22.vcf.gz\t0.002420 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr3.vcf.gz\t0.009032 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr4.vcf.gz\t0.008487 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr5.vcf.gz\t0.008787 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr6.vcf.gz\t0.010675 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr7.vcf.gz\t0.007866 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr8.vcf.gz\t0.009242 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/brain_aging/phase2/genotypes/aging_phase2.hg38.chr9.vcf.gz\t0.008142 seconds\n"
     ]
    }
   ],
   "source": [
    "vcf_files = []\n",
    "for chrom in autosomes:\n",
    "    vcf_files.append(f'{bfile_set}.hg38.chr{chrom}.vcf.gz')\n",
    "\n",
    "vcf_files_arg = ' '.join(vcf_files)\n",
    "this_cmd = f'bcftools concat --output-type z --output {demuxlet_vcf_file} \\\n",
    "--threads {cpu_count} --no-version {vcf_files_arg}'   \n",
    "\n",
    "run_bash_command(this_cmd, DEBUG)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cc6cf61",
   "metadata": {},
   "source": [
    "#### index the final vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "81a318f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!tabix --preset vcf {demuxlet_vcf_file}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db513299-456d-4f58-85f1-70096c0a018a",
   "metadata": {},
   "source": [
    "### split vcfs per pool\n",
    "ie split by vcfs by samples with many extra samples in the vcf demuxlet tends to end up under-assigning singlets, to me\n",
    "\n",
    "smallish number of samples per pool, six, so can specify them on bcftools cmdline\n",
    "\n",
    "need to force and ID update in for sample Aging134, both Aging104 and Aging134 and from same donor so duplicate but genotypes are currently labelled as Aging104; but since separate samples from same donor need to correct the sample ID, effects Pool6 for GEX and ATAC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0df9bef-a96f-4591-bffb-b1a551c97578",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### need to force and ID update in for sample Aging134, both Aging104 and Aging134 and from same donor so duplicate but genotypes are currently labelled as Aging104; but since separate samples from same donor need to correct the sample ID, effects Pool6 for GEX and ATAC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "656e6927-7a70-4112-863a-cb95b4d50088",
   "metadata": {},
   "source": [
    "#### format expected pool names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c1492577-c559-4dbb-a564-0a675213d7ba",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of expected pools (53, 3)\n"
     ]
    }
   ],
   "source": [
    "pool_names = []\n",
    "pools = []\n",
    "lanes = []\n",
    "for pool in info_df.gex_pool.unique():\n",
    "    for lane in lane_range:\n",
    "        gex_sample = f'{src_dir}/gex/sample_ec_GEX_P{pool}_{lane}'\n",
    "        if exists(gex_sample):\n",
    "            pool_names.append(f'GEX_P{pool}_{lane}')\n",
    "            pools.append(pool)\n",
    "            lanes.append(lane)\n",
    "for pool in info_df.atac_pool.unique():\n",
    "    for lane in lane_range:\n",
    "        atac_sample = f'{src_dir}/atac/sample_ec_ATAC_P{pool}_{lane}'\n",
    "        if exists(atac_sample):\n",
    "            pool_names.append(f'ATAC_P{pool}_{lane}')\n",
    "            pools.append(pool)\n",
    "            lanes.append(lane)            \n",
    "this_data = {'name': pool_names, 'pool': pools, 'lane': lanes}\n",
    "pools_df = DataFrame(data=this_data)\n",
    "print(f'shape of expected pools {pools_df.shape}')\n",
    "if DEBUG:\n",
    "    display(pools_df.head())     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "098c8008-2978-45a1-8575-b3f8e038624b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of pools defined = (53, 3)\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "Warn: subset called for sample that does not exist in header: \"Aging142\"... skipping\n",
      "Warn: subset called for sample that does not exist in header: \"Aging142\"... skipping\n",
      "Warn: subset called for sample that does not exist in header: \"Aging142\"... skipping\n",
      "Warn: subset called for sample that does not exist in header: \"Aging142\"... skipping\n",
      "Warn: subset called for sample that does not exist in header: \"Aging142\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "['Aging130' 'Aging131' 'Aging132' 'Aging133' 'Aging134' 'Aging135'\n",
      " 'Aging138' 'Aging139' 'Aging104']\n",
      "Warn: subset called for sample that does not exist in header: \"Aging134\"... skipping\n",
      "Warn: subset called for sample that does not exist in header: \"Aging142\"... skipping\n",
      "Warn: subset called for sample that does not exist in header: \"Aging142\"... skipping\n",
      "Warn: subset called for sample that does not exist in header: \"Aging142\"... skipping\n",
      "Warn: subset called for sample that does not exist in header: \"Aging142\"... skipping\n"
     ]
    }
   ],
   "source": [
    "print(f'number of pools defined = {pools_df.shape}')\n",
    "affected_pools = pools_df.loc[(pools_df.pool == 6)]\n",
    "for row in pools_df.itertuples():\n",
    "    if row.name.startswith('GEX'):\n",
    "        pool_samples = info_df.loc[info_df.gex_pool == row.pool].sample_id.values\n",
    "    elif row.name.startswith('ATAC'):\n",
    "        pool_samples = info_df.loc[info_df.atac_pool == row.pool].sample_id.values\n",
    "    if row.name in affected_pools.name.values:\n",
    "        pool_samples = append(pool_samples, 'Aging104')\n",
    "    fmt_samples_arg = ','.join(pool_samples)\n",
    "    out_vcf = f'{genos_dir}/{proj_name}.hg38.demuxlet.{row.name}.vcf.gz'\n",
    "    this_cmd = f'bcftools view --samples {fmt_samples_arg} --force-samples \\\n",
    "--output {out_vcf} --output-type z {demuxlet_vcf_file} --threads {cpu_count}'\n",
    "    run_bash_command(this_cmd, DEBUG)\n",
    "    # index the pool vcf\n",
    "    run_bash_command(f'tabix --preset vcf {out_vcf}', DEBUG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2e31a676-e38d-49a2-85fb-36400f01afc6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# create a rename file for bcftools to use\n",
    "pool6_rename_file = f'{genos_dir}/P6.duplicate.rename.list'\n",
    "this_cmd = f'echo \"Aging104\\tAging134\" > {pool6_rename_file}'\n",
    "!{this_cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "41ea4e52-2e8e-4184-9c10-ab237138deb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "for pool in affected_pools.name:\n",
    "    in_vcf = f'{genos_dir}/{proj_name}.hg38.demuxlet.{pool}.vcf.gz'\n",
    "    temp_vcf = f'{genos_dir}/{proj_name}.hg38.demuxlet.{pool}.temp.vcf.gz'\n",
    "    this_cmd = f'mv {in_vcf} {temp_vcf}'\n",
    "    run_bash_command(this_cmd, DEBUG)\n",
    "    this_cmd = f'bcftools reheader --sample {pool6_rename_file} \\\n",
    "    --output {in_vcf} --threads 2 {temp_vcf}'\n",
    "    run_bash_command(this_cmd, DEBUG)\n",
    "    # index the pool vcf\n",
    "    run_bash_command(f'tabix -f --preset vcf {out_vcf}', DEBUG)\n",
    "    # remove temp file\n",
    "    run_bash_command(f'rm {temp_vcf}', DEBUG)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1f4a692f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Aug 11 16:08:35 EDT 2023\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
