{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c26f321c",
   "metadata": {},
   "source": [
    "## Notebook for prepping the ADRD Aging genotypes for NIMH Human Brain Collection Core\n",
    "original 3 Illumina chip types ped/map files from [here](gs://nihnialng-aging-brain/genotypes); H1M, H5M4, M650K\n",
    "\n",
    "- H1M=Human1M-Duov3_B (3 samples)\n",
    "- H650K=HumanHap650Yv3.0 (7 samples)\n",
    "- H5M4=HumanOmni5-Quad (3 samples)\n",
    "\n",
    "basically need to:\n",
    "- merge genotypes from different Illumina platforms\n",
    "- liftover from hg19 to hg38\n",
    "- re-order the chromosomes from typical to 10X's lexigraphical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2d9f1fe9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Jul 16 21:36:37 EDT 2021\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc23a89e",
   "metadata": {},
   "source": [
    "#### import libraries and set notebook variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1c144911",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "17c2f586",
   "metadata": {},
   "outputs": [],
   "source": [
    "# naming\n",
    "project = 'adrd'\n",
    "cohort = 'aging'\n",
    "bank = 'nhbcc'\n",
    "\n",
    "# directories\n",
    "wrk_dir = '/labshare/raph/datasets/adrd_neuro'\n",
    "cohort_dir = f'{wrk_dir}/{cohort}'\n",
    "cohort_bckt = 'gs://nihnialng-aging-brain'\n",
    "genos_dir = f'{cohort_dir}/genotypes'\n",
    "genos_bcket = f'{cohort_bckt}/genotypes'\n",
    "tools_dir = f'{wrk_dir}/tools'\n",
    "fasta_index_bucket_path = 'gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta*'\n",
    "fasta_dict_bucket_path = 'gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dict'\n",
    "demuxlet_vcf_file = f'{genos_dir}/{project}_{cohort}_{bank}.hg38.demuxlet.vcf.gz'\n",
    "\n",
    "# constants\n",
    "capture_out = !(nproc)\n",
    "max_threads = int(capture_out[0])\n",
    "capture_out = !grep MemTotal /proc/meminfo | awk '{print $2}'\n",
    "max_mem = int(capture_out[0])\n",
    "\n",
    "hbcc_ped_prefixes = ['H1M', 'H5M4', 'H650K']\n",
    "\n",
    "hbcc_id_dict = {'200901070006_R03C01': 'NHBCC-2790', '204321360038_R01C01': 'NHBCC-2628', \n",
    "                '3999495136_R02C01': 'NHBCC-1119', '4031091001_A': 'NHBCC-1613', \n",
    "                '4031091029_A': 'NHBCC-1615', '4031091062_A': 'NHBCC-1669', \n",
    "                '4040296037_A': 'NHBCC-1556', '4040296068_A': 'NHBCC-1604', \n",
    "                '4040296088_A': 'NHBCC-1137', '4256126251_A': 'NHBCC-1187', \n",
    "                '4463344637_R01C01': 'NHBCC-1275', '4572348358_R01C02': 'NHBCC-831', \n",
    "                '4584664028_R01C01': 'NHBCC-1340'}\n",
    "\n",
    "# when combining final vcf for demuxlet need to have chromosomes \n",
    "#sorted to match 10x lexigraphical\n",
    "autosomes = [1, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 2, 20, 21, 22, 3, 4, 5, 6, 7, 8, 9]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9fc7a63",
   "metadata": {},
   "source": [
    "#### pull down the NIMH HBCC provided plink ped files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9233770b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H1M: gsutil -mq cp gs://nihnialng-aging-brain/genotypes/H1M.* /labshare/raph/datasets/adrd_neuro/aging/genotypes/\n",
      "H5M4: gsutil -mq cp gs://nihnialng-aging-brain/genotypes/H5M4.* /labshare/raph/datasets/adrd_neuro/aging/genotypes/\n",
      "H650K: gsutil -mq cp gs://nihnialng-aging-brain/genotypes/H650K.* /labshare/raph/datasets/adrd_neuro/aging/genotypes/\n"
     ]
    }
   ],
   "source": [
    "for hbcc_prefix in hbcc_ped_prefixes:\n",
    "    this_cmd = f'gsutil -mq cp {genos_bcket}/{hbcc_prefix}.* {genos_dir}/'\n",
    "    print(f'{hbcc_prefix}: {this_cmd}')\n",
    "    !{this_cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a78067bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 333M\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  35M Jul 16 21:36 H1M.map\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  14M Jul 16 21:36 H1M.ped\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 130M Jul 16 21:36 H5M4.map\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  51M Jul 16 21:36 H5M4.ped\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  20M Jul 16 21:36 H650K.map\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  18M Jul 16 21:36 H650K.ped\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  234 Jul 16 13:47 merge-list.txt\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 4.7M Jul 16 13:50 variants_to_keep.txt\n"
     ]
    }
   ],
   "source": [
    "!ls -lh {genos_dir}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b5f48b9",
   "metadata": {},
   "source": [
    "#### see what variants are shared between platforms provided, by variant name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fc47125d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H1M shape=(1192666, 4) shared variant size = 1192666\n",
      "H5M4 shape=(4437269, 4) shared variant size = 971362\n",
      "H650K shape=(660918, 4) shared variant size = 561035\n"
     ]
    }
   ],
   "source": [
    "shared_variants = None\n",
    "for hbcc_prefix in hbcc_ped_prefixes:\n",
    "    map_df = pd.read_csv(f'{genos_dir}/{hbcc_prefix}.map', sep='\\s+', header=None)\n",
    "    # in map file variants name is 2nd column\n",
    "#     print(map_df.shape)\n",
    "#     display(map_df.head())\n",
    "    if shared_variants is None:\n",
    "        shared_variants = set(map_df[1])\n",
    "    else:\n",
    "        shared_variants = shared_variants & set(map_df[1])\n",
    "    print(f'{hbcc_prefix} shape={map_df.shape} shared variant size = {len(shared_variants)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b0a72cc",
   "metadata": {},
   "source": [
    "#### looks like a definitely a decent number of variants shared, proceed to merging"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "343c6e41",
   "metadata": {},
   "source": [
    "#### convert to bfiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "34f59700",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plink --ped /labshare/raph/datasets/adrd_neuro/aging/genotypes/H1M.ped --map /labshare/raph/datasets/adrd_neuro/aging/genotypes/H1M.map --make-bed --out /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc_H1M --silent\n",
      "Warning: 212 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc_H1M.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "plink --ped /labshare/raph/datasets/adrd_neuro/aging/genotypes/H5M4.ped --map /labshare/raph/datasets/adrd_neuro/aging/genotypes/H5M4.map --make-bed --out /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc_H5M4 --silent\n",
      "Warning: 3126 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc_H5M4.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "plink --ped /labshare/raph/datasets/adrd_neuro/aging/genotypes/H650K.ped --map /labshare/raph/datasets/adrd_neuro/aging/genotypes/H650K.map --make-bed --out /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc_H650K --silent\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n"
     ]
    }
   ],
   "source": [
    "for hbcc_prefix in hbcc_ped_prefixes:\n",
    "    this_cmd = f'plink --ped {genos_dir}/{hbcc_prefix}.ped \\\n",
    "--map {genos_dir}/{hbcc_prefix}.map \\\n",
    "--make-bed --out {genos_dir}/{project}_{cohort}_{bank}_{hbcc_prefix} --silent'\n",
    "    print(this_cmd)\n",
    "    !{this_cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a8b3eef7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 596M\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  234 Jul 16 13:47 merge-list.txt\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 4.7M Jul 16 13:50 variants_to_keep.txt\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  14M Jul 16 21:36 H1M.ped\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  35M Jul 16 21:36 H1M.map\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  51M Jul 16 21:36 H5M4.ped\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 130M Jul 16 21:36 H5M4.map\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  18M Jul 16 21:36 H650K.ped\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  20M Jul 16 21:36 H650K.map\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 6.7K Jul 16 21:37 adrd_aging_nhbcc_H1M.hh\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 1.2M Jul 16 21:37 adrd_aging_nhbcc_H1M.bed\n",
      "-rw-rw-r--. 1 gibbsr gibbsr   92 Jul 16 21:37 adrd_aging_nhbcc_H1M.fam\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  40M Jul 16 21:37 adrd_aging_nhbcc_H1M.bim\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 1.8K Jul 16 21:37 adrd_aging_nhbcc_H1M.log\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 107K Jul 16 21:37 adrd_aging_nhbcc_H5M4.hh\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 4.3M Jul 16 21:37 adrd_aging_nhbcc_H5M4.bed\n",
      "-rw-rw-r--. 1 gibbsr gibbsr   95 Jul 16 21:37 adrd_aging_nhbcc_H5M4.fam\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 146M Jul 16 21:37 adrd_aging_nhbcc_H5M4.bim\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 1.8K Jul 16 21:37 adrd_aging_nhbcc_H5M4.log\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  113 Jul 16 21:37 adrd_aging_nhbcc_H650K.nosex\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 1.3M Jul 16 21:37 adrd_aging_nhbcc_H650K.bed\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  176 Jul 16 21:37 adrd_aging_nhbcc_H650K.fam\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  22M Jul 16 21:37 adrd_aging_nhbcc_H650K.bim\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 1.8K Jul 16 21:37 adrd_aging_nhbcc_H650K.log\n"
     ]
    }
   ],
   "source": [
    "!ls -lhtr {genos_dir}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "376bf1a9",
   "metadata": {},
   "source": [
    "#### merge the per Illumina chip type for the plink bfiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "71d455e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plink --merge-list /labshare/raph/datasets/adrd_neuro/aging/genotypes/merge-list.txt --make-bed --allow-no-sex --keep-allele-order --silent --out /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc\n",
      "Warning: Multiple chromosomes seen for variant 'rs10402666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1049702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1131675'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11551867'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11799112'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11925488'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1224751'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1640283'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17883004'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1851319'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1864011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2005417'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2090052'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs252585'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2555849'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2622254'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2635512'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs281377'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28539253'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28693514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2957297'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3212680'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34278206'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35437602'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs381494'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4797621'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs496572'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5945676'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6830714'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs685438'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7212153'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7255768'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs731137'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7600058'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8031106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9647005'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9778500'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9785726'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786108'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786272'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA018897'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7530935'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10182914'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9752952'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA020826'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA022478'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs909274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13154594'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10038870'.\n",
      "Warning: Multiple positions seen for variant 'rs10478694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10807011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35882598'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10260309'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2261257'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3915411'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34499424'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28580083'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA027818'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA011118'.\n",
      "Warning: Multiple positions seen for variant 'rs35661597'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17212015'.\n",
      "Warning: Multiple positions seen for variant 'rs9561064'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs380804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12597978'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2690904'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7267880'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821958'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2334639'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs140281'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13056641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5752647'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2771839'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2568759'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35858670'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34730793'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3102685'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2562942'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2163209'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9782855'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786212'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35740538'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5941215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2552960'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7885800'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2481225'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2437650'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35324435'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34262498'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9657839'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2556924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2565525'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2751196'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4028872'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2435738'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5985048'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2933080'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1774963'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17129789'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs315041'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4364877'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17101527'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2655760'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4649592'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12043679'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1079109'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12128686'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6427609'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2376852'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2754471'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2351422'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2348472'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2339752'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12619064'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2176413'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2192400'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6546667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10200219'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6710843'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12615215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1498957'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6421783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2375763'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs163667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1519514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13425421'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13002544'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2342694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11717239'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9871790'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2028240'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7624950'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918537'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918531'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11706032'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs536575'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3009033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918515'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13326562'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9871837'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1568292'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7628649'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2895461'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2063300'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2279029'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3860706'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1983217'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2291715'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9761692'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2662733'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3925117'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4276244'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10015934'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13120250'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11738302'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13154594'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3903000'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12523271'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11738629'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2569201'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1826532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4959515'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs719065'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596790'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596796'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1535053'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12660307'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2158556'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6937363'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9328044'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2281351'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17133610'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10155688'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1316662'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596815'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9501714'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12191783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6915646'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12192106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6916246'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6913464'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9502915'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9405480'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9502941'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1080100'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1610415'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1045074'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9264536'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs602875'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12660719'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040035'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11760174'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12183238'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1112237'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11759626'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2056474'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10245070'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10224134'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2257576'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs38642'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17152154'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2456554'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17863175'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361379'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163396'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361342'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2252241'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163403'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2213198'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2213199'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13233002'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361377'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11535222'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855945'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2734189'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17282'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16622'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17708955'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855950'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855951'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2734151'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855150'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855954'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12672113'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6947359'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10487532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12534123'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13223129'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1006094'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6979080'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12533751'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1006093'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17257'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17766702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163497'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361473'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7787291'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17767132'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2269332'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17286'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10231034'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12667623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10215320'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040350'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040351'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7780065'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7789642'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855983'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2680432'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1126092'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs682400'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2035685'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs974053'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13258887'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2380567'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2338696'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1113706'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1126274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1121947'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4976940'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7015131'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6421025'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11786011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1446931'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1446934'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7819587'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1470514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11781564'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13291255'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1112042'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10908023'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1578263'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10123966'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2282199'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13285529'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs715666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2026726'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7041980'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11103281'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7025716'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4842084'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs884559'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4842090'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1073731'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1751274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs564166'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10508547'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2437258'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs691461'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2461180'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2813267'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2394498'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10881601'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1158744'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10836370'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2901810'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs714215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4244818'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11820113'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3855349'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10502267'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1975930'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs770987'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1296265'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1278610'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7131737'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11611532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs710944'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7306551'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs770991'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758919'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2272367'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2036288'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7300570'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12422489'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758929'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11611755'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7304881'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7954265'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7979175'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3825108'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7977661'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3847667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7310221'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1587447'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6560914'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758928'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7970251'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7961667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7961682'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs905227'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3805728'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7958575'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11297'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2259444'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11612324'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs595241'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3749826'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2088683'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1924627'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2897385'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2038897'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7984879'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6492347'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11159741'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1954089'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12442023'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2665033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12591365'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2871771'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs438702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1685119'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2858016'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1104612'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4985430'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4788728'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1609516'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1973344'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9708471'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs883635'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs900768'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9284357'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1079841'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3907609'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4636966'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3999623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12939215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4968723'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8069746'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8069177'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4968624'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2070782'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11079538'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7212440'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8066263'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9912957'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8065316'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7503015'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7503902'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9907512'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8080666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7210806'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9899114'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9894547'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7501783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7405659'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8071835'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9889615'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9896532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1619323'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2318319'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs918442'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10414884'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3897509'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8130901'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3855691'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821973'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821997'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2792379'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2775862'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17004452'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2258197'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2822124'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs743067'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1985740'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2258691'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2775054'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11700583'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17001860'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17728665'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361973'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2845371'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5993924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10084637'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2190742'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748614'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748622'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9605145'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9605146'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs759235'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2108585'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16981694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9606468'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748636'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7290901'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5752043'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11912332'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13056641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16979433'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5760804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7286373'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11705626'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs713696'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8142181'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5760813'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs742169'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5996821'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6004408'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2064139'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6005641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1883472'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6519730'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs882455'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs882456'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820778'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6005657'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1004236'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs722343'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11090516'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1062731'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1997739'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs730647'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs14115'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2413140'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5754192'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4821082'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820087'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788477'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12166565'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788479'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5994595'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740985'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740987'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740991'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs114682'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740982'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs130743'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs470033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820090'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1159220'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5998577'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4821086'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788483'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs713867'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs136183'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3021268'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11545763'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2061106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5918554'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6619104'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12557859'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6418513'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6627746'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5925325'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7058908'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5924804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2097253'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17254754'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC464T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA829G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC2791T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC3595T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT4562C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT6153C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA6753G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoG8270A'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT8278C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC8656T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT11486C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT14799C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA15245G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoG16392A'.\n",
      "653506 more multiple-position warnings: see log file.\n",
      "Error: 1 variant with 3+ alleles present.\n",
      "* If you believe this is due to strand inconsistency, try --flip with\n",
      "  /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc-merge.missnp.\n",
      "  (Warning: if this seems to work, strand errors involving SNPs with A/T or C/G\n",
      "  alleles probably remain in your data.  If LD between nearby SNPs is high,\n",
      "  --flip-scan should detect them.)\n",
      "* If you are dealing with genuine multiallelic variants, we recommend exporting\n",
      "  that subset of the data to VCF (via e.g. '--recode vcf'), merging with\n",
      "  another tool/script, and then importing the result; PLINK is not yet suited\n",
      "  to handling them.\n",
      "See https://www.cog-genomics.org/plink/1.9/data#merge3 for more discussion.\n",
      "Warning: 212 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc_H1M.temp.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: 3126 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc_H5M4.temp.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10402666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1049702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1131675'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11551867'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11799112'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11925488'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1224751'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1640283'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17883004'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1851319'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1864011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2005417'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2090052'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs252585'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2555849'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2622254'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2635512'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs281377'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28539253'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28693514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2957297'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3212680'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34278206'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35437602'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs381494'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4797621'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs496572'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5945676'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6830714'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs685438'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7212153'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7255768'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs731137'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7600058'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8031106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9647005'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9778500'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9785726'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786108'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786272'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA018897'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7530935'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10182914'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9752952'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA020826'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA022478'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs909274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13154594'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10038870'.\n",
      "Warning: Multiple positions seen for variant 'rs10478694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10807011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35882598'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10260309'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2261257'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3915411'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34499424'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs28580083'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA027818'.\n",
      "Warning: Multiple chromosomes seen for variant 'GA011118'.\n",
      "Warning: Multiple positions seen for variant 'rs35661597'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17212015'.\n",
      "Warning: Multiple positions seen for variant 'rs9561064'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs380804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12597978'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2690904'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7267880'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821958'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2334639'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs140281'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13056641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5752647'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2771839'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2568759'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35858670'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34730793'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3102685'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2562942'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2163209'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9782855'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9786212'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35740538'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5941215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2552960'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7885800'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2481225'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2437650'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs35324435'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs34262498'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9657839'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2556924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2565525'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2751196'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4028872'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2435738'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5985048'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2933080'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1774963'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17129789'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs315041'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4364877'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17101527'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2655760'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4649592'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12043679'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1079109'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12128686'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6427609'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2376852'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2754471'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2351422'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2348472'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2339752'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12619064'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2176413'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2192400'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6546667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10200219'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6710843'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12615215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1498957'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6421783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2375763'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs163667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1519514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13425421'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13002544'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2342694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11717239'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9871790'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2028240'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7624950'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918537'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918531'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11706032'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs536575'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3009033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2918515'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13326562'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9871837'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1568292'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7628649'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2895461'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2063300'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2279029'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3860706'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1983217'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2291715'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9761692'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2662733'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3925117'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4276244'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10015934'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13120250'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11738302'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13154594'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3903000'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12523271'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11738629'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2569201'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1826532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4959515'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs719065'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596790'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596796'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1535053'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12660307'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2158556'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6937363'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9328044'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2281351'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17133610'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10155688'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1316662'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6596815'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9501714'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12191783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6915646'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12192106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6916246'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6913464'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9502915'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9405480'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9502941'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1080100'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1610415'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1045074'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9264536'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs602875'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12660719'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040035'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11760174'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12183238'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1112237'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11759626'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2056474'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10245070'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10224134'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2257576'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs38642'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17152154'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2456554'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17863175'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361379'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163396'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361342'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2252241'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163403'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2213198'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2213199'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13233002'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361377'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11535222'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855945'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2734189'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17282'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16622'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17708955'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855950'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855951'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2734151'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855150'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855954'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12672113'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6947359'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10487532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12534123'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13223129'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1006094'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6979080'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12533751'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1006093'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17257'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17766702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17163497'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361473'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7787291'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17767132'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2269332'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17286'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10231034'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12667623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10215320'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040350'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2040351'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7780065'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7789642'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2855983'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2680432'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1126092'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs682400'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2035685'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs974053'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13258887'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2380567'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2338696'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1113706'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1126274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1121947'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4976940'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7015131'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6421025'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11786011'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1446931'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1446934'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7819587'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1470514'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11781564'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13291255'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1112042'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10908023'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1578263'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10123966'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2282199'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13285529'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs715666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2026726'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7041980'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11103281'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7025716'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4842084'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs884559'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4842090'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1073731'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1751274'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs564166'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10508547'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2437258'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs691461'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2461180'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2813267'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2394498'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10881601'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1158744'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10836370'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2901810'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs714215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4244818'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11820113'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3855349'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10502267'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1975930'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs770987'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1296265'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1278610'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7131737'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11611532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs710944'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7306551'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs770991'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758919'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2272367'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2036288'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7300570'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12422489'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758929'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11611755'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7304881'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7954265'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7979175'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3825108'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7977661'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3847667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7310221'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1587447'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6560914'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4758928'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7970251'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7961667'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7961682'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs905227'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3805728'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7958575'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11297'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2259444'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11612324'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs595241'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3749826'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2088683'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1924627'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2897385'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2038897'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7984879'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6492347'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11159741'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1954089'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12442023'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2665033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12591365'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2871771'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs438702'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1685119'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2858016'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1104612'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4985430'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4788728'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1609516'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1973344'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9708471'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs883635'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs900768'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9284357'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1079841'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3907609'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4636966'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3999623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12939215'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4968723'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8069746'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8069177'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4968624'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2070782'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11079538'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7212440'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8066263'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9912957'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8065316'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7503015'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7503902'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9907512'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8080666'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7210806'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9899114'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9894547'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7501783'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7405659'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8071835'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9889615'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9896532'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1619323'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2318319'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs918442'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10414884'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3897509'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8130901'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3855691'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821973'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2821997'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2792379'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2775862'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17004452'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2258197'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2822124'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs743067'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1985740'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2258691'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2775054'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11700583'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17001860'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17728665'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs361973'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2845371'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5993924'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs10084637'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2190742'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748614'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748622'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9605145'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748623'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9605146'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs759235'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2108585'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16981694'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs9606468'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5748636'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7290901'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5752043'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11912332'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs13056641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs16979433'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5760804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7286373'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11705626'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs713696'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs8142181'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5760813'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs742169'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5996821'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6004408'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2064139'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6005641'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1883472'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6519730'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs882455'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs882456'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820778'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6005657'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1004236'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs722343'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11090516'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1062731'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1997739'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs730647'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs14115'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2413140'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5754192'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4821082'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820087'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788477'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12166565'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788479'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5994595'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740985'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740987'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740991'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs114682'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2740982'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs130743'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs470033'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4820090'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs1159220'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5998577'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs4821086'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3788483'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs713867'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs136183'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs3021268'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs11545763'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2061106'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5918554'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6619104'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs12557859'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6418513'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs6627746'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5925325'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs7058908'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs5924804'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs2097253'.\n",
      "Warning: Multiple chromosomes seen for variant 'rs17254754'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC464T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA829G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC2791T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC3595T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT4562C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT6153C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA6753G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoG8270A'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT8278C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoC8656T'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT11486C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoT14799C'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoA15245G'.\n",
      "Warning: Multiple chromosomes seen for variant 'MitoG16392A'.\n",
      "653505 more multiple-position warnings: see log file.\n",
      "Warning: Variants 'rs4040617' and 'kgp5225889' have the same position.\n",
      "Warning: Variants 'rs17160939' and 'kgp15219107' have the same position.\n",
      "Warning: Variants 'rs28609852' and 'kgp3324955' have the same position.\n",
      "207087 more same-position warnings: see log file.\n",
      "Warning: 3336 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n"
     ]
    }
   ],
   "source": [
    "# merge the files into a single plink binary set\n",
    "merge_file_set = f'{genos_dir}/merge-list.txt'\n",
    "bfile_set = f'{genos_dir}/{project}_{cohort}_{bank}'\n",
    "\n",
    "with open(merge_file_set, 'w') as file_handler:\n",
    "    for hbcc_prefix in hbcc_ped_prefixes:\n",
    "        prefix_file_set = f'{genos_dir}/{project}_{cohort}_{bank}_{hbcc_prefix}'\n",
    "        file_handler.write(f'{prefix_file_set}\\n')\n",
    "\n",
    "# merge the per platform bfiles into a merged bfile\n",
    "this_cmd = f'plink --merge-list {merge_file_set} --make-bed --allow-no-sex \\\n",
    "--keep-allele-order --silent --out {bfile_set}'\n",
    "print(this_cmd)\n",
    "!{this_cmd}\n",
    "\n",
    "# if there was a missnp problem remove those variant and re-attemp merge\n",
    "if os.path.exists(f'{bfile_set}-merge.missnp'):\n",
    "#     for chrom in CHROMOSOMES:\n",
    "    for hbcc_prefix in hbcc_ped_prefixes:\n",
    "        !plink --bfile {genos_dir}/{project}_{cohort}_{bank}_{hbcc_prefix} \\\n",
    "--silent --exclude {bfile_set}-merge.missnp \\\n",
    "--keep-allele-order --make-bed \\\n",
    "--out {genos_dir}/{project}_{cohort}_{bank}_{hbcc_prefix}.temp\n",
    "\n",
    "    with open(merge_file_set, 'w') as file_handler:\n",
    "        for hbcc_prefix in hbcc_ped_prefixes:\n",
    "            prefix_file_set = f'{genos_dir}/{project}_{cohort}_{bank}_{hbcc_prefix}.temp'\n",
    "            file_handler.write(f'{prefix_file_set}\\n')\n",
    "        \n",
    "    !plink --merge-list {merge_file_set} --make-bed --allow-no-sex \\\n",
    "--keep-allele-order --silent --out {bfile_set} --geno 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a87243ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r--. 1 gibbsr gibbsr  113 Jul 16 21:38 /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.nosex\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 113K Jul 16 21:38 /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hh\n",
      "-rw-rw-r--. 1 gibbsr gibbsr 2.0M Jul 16 21:38 /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.bed\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  363 Jul 16 21:38 /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.fam\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  17M Jul 16 21:38 /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.bim\n",
      "-rw-rw-r--. 1 gibbsr gibbsr  51M Jul 16 21:38 /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.log\n"
     ]
    }
   ],
   "source": [
    "!ls -lhtr {genos_dir}/{project}_{cohort}_{bank}.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "de70d3a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc-merge.bim +\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc-merge.fam .\n",
      "4665423 variants loaded from .bim file.\n",
      "13 people (4 males, 2 females, 7 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 13 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... done.\n",
      "Warning: 3336 het. haploid genotypes present (see\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.344533.\n",
      "4155982 variants removed due to missing genotype data (--geno).\n",
      "509441 variants and 13 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.bed +\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.bim +\n",
      "/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.fam ...\n",
      "done.\n",
      "\n",
      "End time: Fri Jul 16 21:38:31 2021\n"
     ]
    }
   ],
   "source": [
    "!tail -n 25 {genos_dir}/{project}_{cohort}_{bank}.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2e8d51a",
   "metadata": {},
   "source": [
    "#### load the bim files see chrom counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d6ebcb4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(509441, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>200057</th>\n",
       "      <td>6</td>\n",
       "      <td>rs17051952</td>\n",
       "      <td>125.0280</td>\n",
       "      <td>124857859</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434501</th>\n",
       "      <td>17</td>\n",
       "      <td>rs12951306</td>\n",
       "      <td>20.0268</td>\n",
       "      <td>7020507</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375133</th>\n",
       "      <td>13</td>\n",
       "      <td>rs7983863</td>\n",
       "      <td>57.6872</td>\n",
       "      <td>59768722</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359722</th>\n",
       "      <td>12</td>\n",
       "      <td>rs11109229</td>\n",
       "      <td>111.1500</td>\n",
       "      <td>98208150</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235964</th>\n",
       "      <td>7</td>\n",
       "      <td>rs13221692</td>\n",
       "      <td>172.2930</td>\n",
       "      <td>153367707</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460968</th>\n",
       "      <td>18</td>\n",
       "      <td>rs8085865</td>\n",
       "      <td>121.7240</td>\n",
       "      <td>76047493</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223076</th>\n",
       "      <td>7</td>\n",
       "      <td>rs6460671</td>\n",
       "      <td>84.5521</td>\n",
       "      <td>71020307</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486174</th>\n",
       "      <td>21</td>\n",
       "      <td>rs2834385</td>\n",
       "      <td>41.7415</td>\n",
       "      <td>35506430</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403148</th>\n",
       "      <td>14</td>\n",
       "      <td>rs7146643</td>\n",
       "      <td>124.8330</td>\n",
       "      <td>105757392</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180933</th>\n",
       "      <td>6</td>\n",
       "      <td>rs642859</td>\n",
       "      <td>37.1041</td>\n",
       "      <td>16796614</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0           1         2          3  4  5\n",
       "200057   6  rs17051952  125.0280  124857859  0  A\n",
       "434501  17  rs12951306   20.0268    7020507  G  A\n",
       "375133  13   rs7983863   57.6872   59768722  A  G\n",
       "359722  12  rs11109229  111.1500   98208150  A  C\n",
       "235964   7  rs13221692  172.2930  153367707  A  C\n",
       "460968  18   rs8085865  121.7240   76047493  G  A\n",
       "223076   7   rs6460671   84.5521   71020307  G  A\n",
       "486174  21   rs2834385   41.7415   35506430  A  G\n",
       "403148  14   rs7146643  124.8330  105757392  G  A\n",
       "180933   6    rs642859   37.1041   16796614  A  G"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "2     41233\n",
       "1     38796\n",
       "3     34364\n",
       "6     33237\n",
       "5     31397\n",
       "4     30507\n",
       "8     28929\n",
       "7     27247\n",
       "10    27091\n",
       "11    24907\n",
       "12    24819\n",
       "9     23941\n",
       "13    19802\n",
       "14    16599\n",
       "18    15727\n",
       "15    15088\n",
       "16    14891\n",
       "20    12937\n",
       "23    12559\n",
       "17    12455\n",
       "19     7605\n",
       "21     7277\n",
       "22     6659\n",
       "24      981\n",
       "0       302\n",
       "26       78\n",
       "25       13\n",
       "Name: 0, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bim_df = pd.read_csv(f'{bfile_set}.bim', sep='\\s+', header=None)\n",
    "print(bim_df.shape)\n",
    "display(bim_df.sample(10))\n",
    "display(bim_df[0].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8879453",
   "metadata": {},
   "source": [
    "#### find variants to exclude that aren't SNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "902b9543",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A    255595\n",
       "G    183262\n",
       "C     43351\n",
       "0     27225\n",
       "T         8\n",
       "Name: 4, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "A    238689\n",
       "G    221471\n",
       "C     49262\n",
       "T        19\n",
       "Name: 5, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(482216, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>rs10015934</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>rs1004236</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>rs1006094</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>rs10084637</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0</td>\n",
       "      <td>rs10155688</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0           1    2  3  4  5\n",
       "1  0  rs10015934  0.0  0  A  G\n",
       "2  0   rs1004236  0.0  0  A  G\n",
       "3  0   rs1006094  0.0  0  A  G\n",
       "4  0  rs10084637  0.0  0  C  A\n",
       "5  0  rs10155688  0.0  0  G  A"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(bim_df[4].value_counts())\n",
    "display(bim_df[5].value_counts())\n",
    "nucleotides = ['A', 'C', 'G', 'T', 'N', 'a', 'c', 'g', 't', 'n', 0]\n",
    "vars_to_include = bim_df.loc[(bim_df[4].isin(nucleotides)) & \n",
    "                             (bim_df[5].isin(nucleotides))]\n",
    "print(vars_to_include.shape)\n",
    "display(vars_to_include.head())\n",
    "\n",
    "vars_to_include[1].to_csv(f'{genos_dir}/variants_to_keep.txt', index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17d8f947",
   "metadata": {},
   "source": [
    "#### convert to vcf, exclude InDels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "90892b07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plink2 --bfile /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc --silent --export vcf-4.2 bgz id-paste=iid --out /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc --output-chr chrM --not-chr 0 --snps-only --extract /labshare/raph/datasets/adrd_neuro/aging/genotypes/variants_to_keep.txt\n",
      "Warning: '_' present in original sample IDs; --export vcf will not be able to\n",
      "reconstruct them. Consider rerunning with a suitable --export id-delim= value.\n"
     ]
    }
   ],
   "source": [
    "this_cmd = f'plink2 --bfile {bfile_set} --silent \\\n",
    "--export vcf-4.2 bgz id-paste=iid --out {bfile_set} \\\n",
    "--output-chr chrM --not-chr 0 --snps-only \\\n",
    "--extract {genos_dir}/variants_to_keep.txt'\n",
    "\n",
    "print(this_cmd)\n",
    "!{this_cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "930d8545",
   "metadata": {},
   "source": [
    "#### re-IDs the NIMH HBCC samples using the provided mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f34709b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>newID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4040296088_A</td>\n",
       "      <td>NHBCC-1137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4040296037_A</td>\n",
       "      <td>NHBCC-1556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>204321360038_R01C01</td>\n",
       "      <td>NHBCC-2628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4584664028_R01C01</td>\n",
       "      <td>NHBCC-1340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>200901070006_R03C01</td>\n",
       "      <td>NHBCC-2790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ID       newID\n",
       "0         4040296088_A  NHBCC-1137\n",
       "1         4040296037_A  NHBCC-1556\n",
       "2  204321360038_R01C01  NHBCC-2628\n",
       "3    4584664028_R01C01  NHBCC-1340\n",
       "4  200901070006_R03C01  NHBCC-2790"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# since using bcftools reheader to update IDs the sample order has to match\n",
    "# check expected samples present and if any ID formats need correction\n",
    "# all files have same sample set so checking one is fine\n",
    "temp_sample_list_file = f'{bfile_set}.sample.list'\n",
    "!bcftools query --list-samples {bfile_set}.vcf.gz > \\\n",
    "{temp_sample_list_file}\n",
    "\n",
    "ids_present_df = pd.read_csv(temp_sample_list_file, header=None)\n",
    "ids_present_df.columns = ['ID']\n",
    "print(ids_present_df.shape)\n",
    "\n",
    "# lookkup the new ID\n",
    "ids_present_df['newID'] = ids_present_df['ID'].apply(hbcc_id_dict.get)\n",
    "\n",
    "ids_present_df.to_csv(f'{bfile_set}.rename.sample.list',\n",
    "                      index=False, header=False, sep='\\t')\n",
    "\n",
    "ids_present_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2b9cf465",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bcftools reheader --sample /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.rename.sample.list --output /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.renamed.vcf.gz --threads 2 /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "in_vcf = f'{bfile_set}.vcf.gz'\n",
    "out_vcf = f'{bfile_set}.renamed.vcf.gz'\n",
    "this_cmd = f'bcftools reheader --sample {bfile_set}.rename.sample.list \\\n",
    "--output {out_vcf} --threads 2 {in_vcf}'\n",
    "print(this_cmd)\n",
    "!{this_cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3d7bd10",
   "metadata": {},
   "source": [
    "#### need to liftover from hg19 to hg38\n",
    "use Picard, slower but better than CrossMap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc40e055",
   "metadata": {},
   "source": [
    "#### get Picard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dd995a0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# grab picard jar\n",
    "!wget --quiet  https://github.com/broadinstitute/picard/releases/download/2.25.5/picard.jar \\\n",
    "-O {tools_dir}/picard.jar"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64da3751",
   "metadata": {},
   "source": [
    "#### grab necessary ref files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "83636ddc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -mq cp -P gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta* /labshare/raph/datasets/adrd_neuro/tools/\n",
      "gsutil -mq cp -P gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dict /labshare/raph/datasets/adrd_neuro/tools/\n"
     ]
    }
   ],
   "source": [
    "this_cmd = f'gsutil -mq cp -P {fasta_index_bucket_path} {tools_dir}/'\n",
    "print(this_cmd)\n",
    "!{this_cmd}\n",
    "\n",
    "this_cmd = f'gsutil -mq cp -P {fasta_dict_bucket_path} {tools_dir}/'\n",
    "print(this_cmd)\n",
    "!{this_cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a202c6f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -mq cp -P gs://gcp-public-data--broad-references/hg19/v0/Homo_sapiens_assembly19.fasta* /labshare/raph/datasets/adrd_neuro/tools/\n",
      "gsutil -mq cp -P gs://gcp-public-data--broad-references/hg19/v0/Homo_sapiens_assembly19.dict /labshare/raph/datasets/adrd_neuro/tools/\n"
     ]
    }
   ],
   "source": [
    "# also need the hg19 files\n",
    "this_cmd = f'gsutil -mq cp -P gs://gcp-public-data--broad-references/hg19/v0/Homo_sapiens_assembly19.fasta* {tools_dir}/'\n",
    "print(this_cmd)\n",
    "!{this_cmd}\n",
    "\n",
    "this_cmd = f'gsutil -mq cp -P gs://gcp-public-data--broad-references/hg19/v0/Homo_sapiens_assembly19.dict {tools_dir}/'\n",
    "print(this_cmd)\n",
    "!{this_cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "666e50b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget --quiet  http://hgdownload.cse.ucsc.edu/gbdb/hg19/liftOver/hg19ToHg38.over.chain.gz \\\n",
    "-O {tools_dir}/hg19ToHg38.over.chain.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1a5c199",
   "metadata": {},
   "source": [
    "#### run the liftover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "71b3a25b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "java -Xmx1056503524k -jar /labshare/raph/datasets/adrd_neuro/tools/picard.jar LiftoverVcf INPUT=/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.renamed.vcf.gz OUTPUT=/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.vcf.gz CHAIN=/labshare/raph/datasets/adrd_neuro/tools/hg19ToHg38.over.chain.gz REJECT=/labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38unmapped.vcf.gz REFERENCE_SEQUENCE=/labshare/raph/datasets/adrd_neuro/tools/Homo_sapiens_assembly38.fasta MAX_RECORDS_IN_RAM=500000 QUIET=true RECOVER_SWAPPED_REF_ALT=true\n",
      "INFO\t2021-07-16 21:54:50\tLiftoverVcf\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    LiftoverVcf -INPUT /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.renamed.vcf.gz -OUTPUT /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.vcf.gz -CHAIN /labshare/raph/datasets/adrd_neuro/tools/hg19ToHg38.over.chain.gz -REJECT /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38unmapped.vcf.gz -REFERENCE_SEQUENCE /labshare/raph/datasets/adrd_neuro/tools/Homo_sapiens_assembly38.fasta -MAX_RECORDS_IN_RAM 500000 -QUIET true -RECOVER_SWAPPED_REF_ALT true\n",
      "**********\n",
      "\n",
      "\n",
      "21:54:50.632 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/labshare/raph/datasets/adrd_neuro/tools/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "INFO\t2021-07-16 21:54:50\tLiftoverVcf\tLoading up the target reference genome.\n",
      "INFO\t2021-07-16 21:54:57\tLiftoverVcf\tLifting variants over and sorting (not yet writing the output file.)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tProcessed 481935 variants.\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\t209 variants failed to liftover.\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\t241244 variants lifted over but had mismatching reference alleles after lift over.\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\t50.1007% of variants were not successfully lifted over and written to the output.\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tliftover success by source contig:\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr1: 18199 / 36679 (49.6169%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr10: 12615 / 25547 (49.3796%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr11: 11785 / 23536 (50.0722%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr12: 11715 / 23443 (49.9723%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr13: 9452 / 18741 (50.4349%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr14: 7856 / 15697 (50.0478%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr15: 7133 / 14331 (49.7732%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr16: 7209 / 14177 (50.8500%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr17: 5966 / 11892 (50.1682%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr18: 7479 / 14992 (49.8866%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr19: 3595 / 7269 (49.4566%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr2: 19498 / 39198 (49.7423%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr20: 6183 / 12319 (50.1908%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr21: 3501 / 6932 (50.5049%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr22: 3216 / 6314 (50.9344%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr3: 16368 / 32667 (50.1056%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr4: 14403 / 29084 (49.5221%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr5: 14882 / 29745 (50.0319%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr6: 15659 / 31261 (50.0912%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr7: 12712 / 25835 (49.2046%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr8: 13844 / 27542 (50.2650%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr9: 11240 / 22671 (49.5788%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchrM: 17 / 29 (58.6207%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchrX: 5774 / 11664 (49.5027%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchrXY: 0 / 11 (0.0000%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchrY: 181 / 359 (50.4178%)\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tlifted variants by target contig:\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr1: 18199\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr10: 12615\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr11: 11785\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr12: 11715\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr13: 9452\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr14: 7856\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr15: 7133\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr16: 7209\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr17: 5966\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr18: 7479\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr19: 3595\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr2: 19498\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr20: 6183\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr21: 3501\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr22: 3216\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr3: 16368\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr4: 14403\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr5: 14882\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr6: 15659\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr7: 12684\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr7_KI270803v1_alt: 28\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr8: 13835\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr8_KI270821v1_alt: 9\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchr9: 11240\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchrM: 17\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchrX: 5774\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tchrY: 181\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\t82409 variants were lifted by swapping REF/ALT alleles.\n",
      "INFO\t2021-07-16 21:55:03\tLiftoverVcf\tWriting out sorted records to final VCF.\n"
     ]
    }
   ],
   "source": [
    "vcf = f'{bfile_set}.renamed.vcf.gz'\n",
    "out_vcf_name = f'{bfile_set}.hg38.vcf.gz'\n",
    "out_vcf_unmapped_name = f'{bfile_set}.hg38unmapped.vcf.gz'\n",
    "\n",
    "this_cmd = \tf'java -Xmx{max_mem}k -jar {tools_dir}/picard.jar LiftoverVcf \\\n",
    "INPUT={vcf} \\\n",
    "OUTPUT={out_vcf_name} \\\n",
    "CHAIN={tools_dir}/hg19ToHg38.over.chain.gz \\\n",
    "REJECT={out_vcf_unmapped_name} \\\n",
    "REFERENCE_SEQUENCE={tools_dir}/Homo_sapiens_assembly38.fasta \\\n",
    "MAX_RECORDS_IN_RAM=500000 QUIET=true RECOVER_SWAPPED_REF_ALT=true'\n",
    "\n",
    "print(this_cmd)\n",
    "!{this_cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b35486a",
   "metadata": {},
   "source": [
    "#### split vcf by chromosome so can be recombined in specified chromosome order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "cb54ba87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.10.11.12.13.14.15.16.17.18.19.2.20.21.22.3.4.5.6.7.8.9.CPU times: user 84 ms, sys: 267 ms, total: 351 ms\n",
      "Wall time: 5.53 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "vcf_files = []\n",
    "for chrom in autosomes:\n",
    "    in_vcf = f'{bfile_set}.hg38.vcf.gz'\n",
    "#     out_vcf = f'{bfile_set}.hg38.chr{chrom}.vcf.gz'\n",
    "    out_vcf = f'{bfile_set}.hg38.chr{chrom}'\n",
    "    # use plink2 instead of bcftools so header is reduced specifically for contigs\n",
    "#     this_cmd = f'bcftools view --threads {max_threads} --output-type z \\\n",
    "# --regions chr{chrom} --output {out_vcf} {in_vcf}'\n",
    "    this_cmd = f'plink2 --vcf {in_vcf} --silent --chr {chrom} --not-chr 0 \\\n",
    "--export vcf-4.2 bgz id-paste=iid --out {out_vcf} --output-chr chrM --allow-extra-chr'\n",
    "    print(chrom, end='.')\n",
    "#     print(this_cmd)\n",
    "    !{this_cmd}    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0f8bddb",
   "metadata": {},
   "source": [
    "#### concat chromosome vcfs (in order) into genome vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "dd78507c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking the headers and starting positions of 22 files\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr1.vcf.gz\t0.010531 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr10.vcf.gz\t0.008038 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr11.vcf.gz\t0.006002 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr12.vcf.gz\t0.005656 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr13.vcf.gz\t0.005266 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr14.vcf.gz\t0.005382 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr15.vcf.gz\t0.003957 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr16.vcf.gz\t0.005047 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr17.vcf.gz\t0.003342 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr18.vcf.gz\t0.004533 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr19.vcf.gz\t0.003049 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr2.vcf.gz\t0.006626 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr20.vcf.gz\t0.004963 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr21.vcf.gz\t0.002895 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr22.vcf.gz\t0.003255 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr3.vcf.gz\t0.008114 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr4.vcf.gz\t0.007416 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr5.vcf.gz\t0.006794 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr6.vcf.gz\t0.005523 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr7.vcf.gz\t0.005459 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr8.vcf.gz\t0.005799 seconds\n",
      "Concatenating /labshare/raph/datasets/adrd_neuro/aging/genotypes/adrd_aging_nhbcc.hg38.chr9.vcf.gz\t0.005965 seconds\n"
     ]
    }
   ],
   "source": [
    "vcf_files = []\n",
    "for chrom in autosomes:\n",
    "    vcf_files.append(f'{bfile_set}.hg38.chr{chrom}.vcf.gz')\n",
    "\n",
    "vcf_files_arg = ' '.join(vcf_files)\n",
    "this_cmd = f'bcftools concat --output-type z --output {demuxlet_vcf_file} \\\n",
    "--threads {max_threads} --no-version {vcf_files_arg}'   \n",
    "\n",
    "# print(this_cmd)\n",
    "!{this_cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cc6cf61",
   "metadata": {},
   "source": [
    "#### index the final vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "81a318f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix --preset vcf {demuxlet_vcf_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f4a692f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
